domain/application/service/resource.go

```diff
@@ -36,34 +38,42 @@ type ResourceState interface {
 	// SetUnitResource sets the resource metadata for a specific unit.
 	SetUnitResource(ctx context.Context, config resource.SetUnitResourceArgs) (resource.SetUnitResourceResult, error)
 
-	// OpenResource returns the metadata for a resource and a reader for
-	// the resource.
-	OpenResource(ctx context.Context, resourceID resources.ID) (resource.Resource, io.ReadCloser, error)
-	// OpenApplicationResource?
+	// OpenApplicationResource returns the metadata for an application's resource.
+	OpenApplicationResource(ctx context.Context, resourceID resources.ID) (resource.Resource, error)
 
-	// OpenResourceForUniter returns the metadata for a resource and a reader
-	// for the resource. A unit resource is created to track the given unit and
-	// which resource its using.
-	OpenResourceForUniter(ctx context.Context, resourceID resources.ID, unitID coreunit.UUID) (resource.Resource, io.ReadCloser, error)
-	// OpenUnitResource?
+	// OpenUnitResource returns the metadata for a resource a. A unit resource is
+	// created to track the given unit and which resource its using.
+	OpenUnitResource(ctx context.Context, resourceID resources.ID, unitID coreunit.UUID) (resource.Resource, error)
 
 	// SetRepositoryResources sets the "polled" resources for the
 	// application to the provided values. The current data for this
 	// application/resource combination will be overwritten.
 	SetRepositoryResources(ctx context.Context, config resource.SetRepositoryResourcesArgs) error
 }
 
+// resourceStoreFn returns the appropriate object store for the
+// provided resource type.
+type resourceStoreFn func(context.Context, charmresource.Type) (objectstore.ObjectStore, error)
```

Comment:
The return type may evolve into a slightly different interface. However we don't currently know what it will be til you write it with the wrapper for writing an oci-image. The type can be updated at that point.

Reply:
With Simon's changes the types have been updated.